src/runtime/internal/ssr.ts

```diff
@@ -32,16 +32,31 @@ export function spread(args, classes_to_add) {
 	return str;
 }
 
-export const escaped = {
+const ATTR_REGEX = /[&"]/g;
+const CONTENT_REGEX = /[&<]/g;
+
+const escapes = {
 	'"': '&quot;',
-	"'": '&#39;',
 	'&': '&amp;',
-	'<': '&lt;',
-	'>': '&gt;'
+	'<': '&lt'
 };
 
-export function escape(html) {
-	return String(html).replace(/["'&<>]/g, match => escaped[match]);
+export function escape(html: string, is_attr = false) {
+	if (typeof html !== 'string') return html;
+
+	const pattern = is_attr ? ATTR_REGEX : CONTENT_REGEX;
+	pattern.lastIndex = 0;
+
+	let escaped = '';
+	let last = 0;
+
+	while (pattern.test(html)) {
+		const i = pattern.lastIndex - 1;
+		escaped += html.slice(last, i) + escapes[html[i]];
```

Comment:
for what it's worth, replacing `escapes[html[i]]` with `const ch = html[i]` and then `(ch === '&' ? ch === '"' ? '&quot;' : '&amp;' : '&lt;')` seems to shave another 10% off the execution time when I test it.

Reply:
Wow. Maybe we should just do that instead then? How are you getting those numbers, and can you share them?

Reply:
I just took [a 500 kb HTML document](https://en.wikipedia.org/wiki/Quantum_mechanics) and ran

```
const html = readFileSync('bigdocument.html').toString();

const start = performance.now();

for (let i = 0; i < 2000; i++) {
  escape(html);
}

console.log(performance.now() - start, 'ms');
```

Our documents of course tend to be smaller so there is a risk this doesn't correspond to real-life performance.

Reply:
The `.replace.replace` version runs in 3.5s,  the code above with `escapres[html[i]]` replaced by the ternary operator in about 3.6, and the original code with a function as second parameter to replace in 14.5s.

Reply:
In my testing a `switch` on character code (`charCodeAt`) is WAY faster than a simple object lookup. Don't ask me why. IE: (from another implementation)

```js
  while (match = matchHtmlRegExp.exec(str)) {
    switch (str.charCodeAt(match.index)) {
      case 34: // "
        escape = '&quot;'
        break
```

Reply:
> html.replace(/&/g, '&amp;').replace(/</g, '&lt;') is pretty much exactly the same speed as the new code.

@ehrencrona In my testing I haven't seen this at all.  Even an unoptimized version of this (with switch) is faster than sequential replace.  The optimized version (using regex to advance the cursor) should be even faster.

```
regex start then one char at time x 355 ops/sec ±0.60% (85 runs sampled)
sequential simple x 304 ops/sec ±0.50% (86 runs sampled)
```

Reply:
>  just submitted a 30% speed improvement patch to replace-html library.

Here's the PR for reference with quite a bit of discussion: https://github.com/component/escape-html/pull/28. It looks like it ended up being closer to 10% over the original implementation in that librar than 30%, but still a nice win. Though it's unclear how it compares to the alternatives here

Reply:
> (ch === '&' ? ch === '"' ? '&quot;' : '&amp;' : '&lt;')

I believe this is broken and it needs to be `(ch === '&' ? '&amp;' : (ch === '"' ? '&quot;' : '&lt;'))`. It does appear faster than the original version in this PR

Reply:
The `switch` method is slower than both the ternary and the original PR implementation in my testing

Reply:
> The switch method is slower than both the ternary and the original PR implementation in my testing

Another thing we have to keep in mind here is that almost 2 years have passed.  It's very hard IMHO to micro-tune JS performance (over the long-term).  Engines vary, browsers vary, things change with time.  The thing that was fastest 2 years ago might bench worse today. The regex engines used in the major browsers are not all created equal.

We should be clear if we're talking benchmarks which browser, which engine... in the Svelt context perhaps we only care about Node?    Just a consideration.

Reply:
Yes, I tested on Node. This is for server-side code, so no need to test in browsers.

Thanks for sharing your findings! It may well have been different in the past like you said and regardless it was valuable to test various  ideas and ensure we're doing as well as possible.

Reply:
Oh for sure... we can certainly chase the peak performance (and probably should)... but we just may have to do so again in another 2 years - and make sure we're always comparing apples to apples. :). If we only need to test on Node that certainly helps a lot.  Though it might not surprise me to learn if there were say differences between Node on  ARM vs Node on x86_64...